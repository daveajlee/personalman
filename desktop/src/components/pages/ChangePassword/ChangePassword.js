import React, {useState} from "react";
import {Button, Col, Container, Form, Image, Row} from "react-bootstrap";
import axios from "axios";
import {useLocation, useNavigate} from "react-router-dom";
import Header from "../../layout/Header/Header";

/**
 * This is the page which allows the user to change their password assuming they known their old password. Otherwise,
 * they must request an admin user to reset their password.
 * @returns {JSX.Element} to be displayed to the user.
 */
function ChangePassword (props) {

    const [currentPassword, setCurrentPassword] = useState("");
    const [newPassword, setNewPassword] = useState("");
    const [confirmedPassword, setConfirmedPassword] = useState("");
    const navigate = useNavigate();
    const location = useLocation();

    /**
     * Retrieve the current token either from the supplied state or empty if we are in doc mode.
     * @returns the current token as a string.
     */
    function getToken() {
        if ( props.docMode && props.docMode==='true') {
            return " - ";
        } else {
            return location.state.token;
        }
    }

    /**
     * Retrieve the company either from the supplied state or empty if we are in doc mode.
     * @returns the current company as a string.
     */
    function getCompany() {
        if ( props.docMode && props.docMode==='true') {
            return "";
        } else {
            return location.state.company;
        }
    }

    /**
     * Set the current password that the user entered to the state for later.
     * @param event the event triggered by the user.
     */
    function currentPasswordChangeHandler(event) {
        setCurrentPassword(event.target.value);
    }

    /**
     * Set the new password that the user entered to the state for later.
     * @param event the event triggered by the user.
     */
    function newPasswordChangeHandler(event) {
        setNewPassword(event.target.value);
    }

    /**
     * Set the confirmed password that the user entered to the state for later.
     * @param event the event triggered by the user.
     */
    function confirmedPasswordChangeHandler(event) {
        setConfirmedPassword(event.target.value)
    }

    /**
     * Method to reset form to initial values.
     */
    function resetForm () {
        setCurrentPassword("");
        setNewPassword("");
        setConfirmedPassword("");
    }

    /**
     *  Change the password by sending the request to the API assuming that the new and confirmed password
     *  are the same. Display an alert with either password changed successfully or error message.
     */
    function changePassword() {
        if ( newPassword !== confirmedPassword ) {
            alert('The two new passwords were not identical! Password could not be changed');
            return;
        }
        axios.patch(process.env.REACT_APP_SERVER_URL + '/user/password', {
            company: getCompany(),
            username: getUsername(),
            token: getToken(),
            currentPassword: currentPassword,
            newPassword: newPassword
        }).then(function (response) {
            if ( response.status === 200 ) {
                alert('Password was changed successfully! Please login again to continue');
                navigate("/")
            }
        }).catch(function (error) {
            alert('Password could not be changed. Please try again later.');
        });
    }

    /**
     * Get the username of the currently logged in user who we will display absences for.
     * @returns the username
     */
    function getUsername() {
        return getToken().split("-")[0];
    }

    /**
     * Display the relevant elements and data to the user.
     */
    return (
        <Container fluid>
            <Header token={getToken()} company={getCompany()}/>
        <Container fluid className="p-3 my-5 h-custom">
        <Row className="d-flex flex-row align-items-center justify-content-center">
            <Col>
                <Image src="https://www.davelee.de/common/assets/img/portfolio/personalman-logo.png"
                       className="d-block mx-auto img-fluid w-50" alt="PersonalMan Logo"/>
            </Col>
            <Col>
                <Container className="d-flex flex-row align-items-center justify-content-center mb-3">
                    <Row>
                        <Col>
                            <h3>Change the password for user: {getUsername()}</h3>
                        </Col>
                    </Row>
                </Container>

                <Container className="d-flex flex-column">
                    <Form>

                        <Form.Group as={Row} className="mb-3" controlId="formPlaintextCurrentPassword">
                            <Form.Label column sm="2">Current Password:</Form.Label>
                            <Col sm="10">
                                <Form.Control value={currentPassword} type="password" placeholder="Current Password" onChange={currentPasswordChangeHandler}/>
                            </Col>
                        </Form.Group>

                        <Form.Group as={Row} className="mb-3" controlId="formPlaintextNewPassword">
                            <Form.Label column sm="2">New Password:</Form.Label>
                            <Col sm="10">
                                <Form.Control value={newPassword} type="password" placeholder="New Password" onChange={newPasswordChangeHandler}/>
                            </Col>
                        </Form.Group>

                        <Form.Group as={Row} className="mb-3" controlId="formPlaintextConfirmPassword">
                            <Form.Label column sm="2">Confirm Password:</Form.Label>
                            <Col sm="10">
                                <Form.Control value={confirmedPassword} type="password" placeholder="Confirmed Password" onChange={confirmedPasswordChangeHandler}/>
                            </Col>
                        </Form.Group>

                    </Form>
                </Container>

                <Container className='align-items-center justify-content-center text-md-start mt-4 pt-2'>
                    <Row>
                        <Col className="text-center">
                            <Button className="mb-0 px-5 me-2" size='lg' onClick={changePassword}>Change Password</Button>
                            <Button className="mb-0 px-5 me-2" size='lg' onClick={resetForm}>Reset</Button>
                        </Col>
                    </Row>
                </Container>
            </Col>
        </Row>
    </Container>
        </Container>)

}

export default ChangePassword;
