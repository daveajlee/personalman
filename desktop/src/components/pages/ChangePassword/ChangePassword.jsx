import React, {useState} from "react";
import {Button, Col, Container, Form, Image, Row} from "react-bootstrap";
import axios from "axios";
import {useLocation, useNavigate} from "react-router-dom";
import Header from "../../layout/Header/Header";
import {useTranslation} from "react-i18next";
import logo from './../../../assets/personalman-logo.png';

/**
 * This is the page which allows the user to change their password assuming they known their old password. Otherwise,
 * they must request an admin user to reset their password.
 * @returns {JSX.Element} to be displayed to the user.
 */
function ChangePassword (props) {

    const [currentPassword, setCurrentPassword] = useState("");
    const [newPassword, setNewPassword] = useState("");
    const [confirmedPassword, setConfirmedPassword] = useState("");
    const navigate = useNavigate();
    const location = useLocation();

    const {t} = useTranslation();

    /**
     * Retrieve the current token either from the supplied state or empty if we are in doc mode.
     * @returns the current token as a string.
     */
    function getToken() {
        if ( props.docMode && props.docMode==='true') {
            return " - ";
        } else {
            return location.state.token;
        }
    }

    /**
     * Retrieve the company either from the supplied state or empty if we are in doc mode.
     * @returns the current company as a string.
     */
    function getCompany() {
        if ( props.docMode && props.docMode==='true') {
            return "";
        } else {
            return location.state.company;
        }
    }

    /**
     * Set the current password that the user entered to the state for later.
     * @param event the event triggered by the user.
     */
    function currentPasswordChangeHandler(event) {
        setCurrentPassword(event.target.value);
    }

    /**
     * Set the new password that the user entered to the state for later.
     * @param event the event triggered by the user.
     */
    function newPasswordChangeHandler(event) {
        setNewPassword(event.target.value);
    }

    /**
     * Set the confirmed password that the user entered to the state for later.
     * @param event the event triggered by the user.
     */
    function confirmedPasswordChangeHandler(event) {
        setConfirmedPassword(event.target.value)
    }

    /**
     * Method to reset form to initial values.
     */
    function resetForm () {
        setCurrentPassword("");
        setNewPassword("");
        setConfirmedPassword("");
    }

    /**
     *  Change the password by sending the request to the API assuming that the new and confirmed password
     *  are the same. Display an alert with either password changed successfully or error message.
     */
    function changePassword() {
        if ( newPassword !== confirmedPassword ) {
            alert(t('changePasswordNotIdentical'));
            return;
        }
        axios.patch(import.meta.env.REACT_APP_SERVER_URL + '/user/password', {
            company: getCompany(),
            username: getUsername(),
            token: getToken(),
            currentPassword: currentPassword,
            newPassword: newPassword
        }).then(function (response) {
            if ( response.status === 200 ) {
                alert(t('changePasswordSuccess'));
                navigate("/")
            }
        }).catch(function (error) {
            alert(t('changePasswordError'));
        });
    }

    /**
     * Get the username of the currently logged in user who we will display absences for.
     * @returns the username
     */
    function getUsername() {
        return getToken().split("-")[0];
    }

    /**
     * Display the relevant elements and data to the user.
     */
    return (
        <Container fluid>
            <Header token={getToken()} company={getCompany()}/>
        <Container fluid className="p-3 my-5 h-custom">
        <Row className="d-flex flex-row align-items-center justify-content-center">
            <Col>
                <Image src={logo}
                       className="d-block mx-auto img-fluid w-50" alt="PersonalMan Logo"/>
            </Col>
            <Col>
                <Container className="d-flex flex-row align-items-center justify-content-center mb-3">
                    <Row>
                        <Col>
                            <h3>{t('changePasswordTitle')} {getUsername()}</h3>
                        </Col>
                    </Row>
                </Container>

                <Container className="d-flex flex-column">
                    <Form>

                        <Form.Group as={Row} className="mb-3" controlId="formPlaintextCurrentPassword">
                            <Form.Label column sm="2">{t('changePasswordCurrentPassword')}:</Form.Label>
                            <Col sm="10">
                                <Form.Control value={currentPassword} type="password" placeholder={t('changePasswordCurrentPassword')} onChange={currentPasswordChangeHandler}/>
                            </Col>
                        </Form.Group>

                        <Form.Group as={Row} className="mb-3" controlId="formPlaintextNewPassword">
                            <Form.Label column sm="2">{t('changePasswordNewPassword')}:</Form.Label>
                            <Col sm="10">
                                <Form.Control value={newPassword} type="password" placeholder={t('changePasswordNewPassword')} onChange={newPasswordChangeHandler}/>
                            </Col>
                        </Form.Group>

                        <Form.Group as={Row} className="mb-3" controlId="formPlaintextConfirmPassword">
                            <Form.Label column sm="2">{t('changePasswordConfirmPassword')}:</Form.Label>
                            <Col sm="10">
                                <Form.Control value={confirmedPassword} type="password" placeholder={t('changePasswordConfirmPassword')} onChange={confirmedPasswordChangeHandler}/>
                            </Col>
                        </Form.Group>

                    </Form>
                </Container>

                <Container className='align-items-center justify-content-center text-md-start mt-4 pt-2'>
                    <Row>
                        <Col className="text-center">
                            <Button className="mb-0 px-5 me-2" size='lg' onClick={changePassword}>{t('changePasswordButton')}</Button>
                            <Button className="mb-0 px-5 me-2" size='lg' onClick={resetForm}>{t('changePasswordReset')}</Button>
                        </Col>
                    </Row>
                </Container>
            </Col>
        </Row>
    </Container>
        </Container>)

}

export default ChangePassword;
