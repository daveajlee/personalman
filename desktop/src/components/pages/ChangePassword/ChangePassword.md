```jsx
import { HashRouter } from "react-router-dom";

<HashRouter>
    <ChangePassword docMode="true"/>
</HashRouter>
```
