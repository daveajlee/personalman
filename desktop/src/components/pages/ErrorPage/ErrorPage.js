import React from "react";
import {Col, Container, Image, Row} from "react-bootstrap";
import logo from './../../../assets/personalman-logo.png';

/**
 * This page is the error page which is displayed whenever the user visits an invalid url.
 * @returns {JSX.Element} to be displayed to the user.
 */
function ErrorPage() {

    /**
     * Display the relevant elements and data to the user.
     */
    return (
        <Container fluid className="p-3 my-5 h-custom">
            <Row className="d-flex flex-row align-items-center justify-content-center">
                <Col>
                    <Image src={logo}
                           className="d-block mx-auto img-fluid w-50" alt="PersonalMan Logo"/>
                </Col>
                <Col>
                    <Container className="d-flex flex-row align-items-center justify-content-center mb-3">
                        <Row>
                            <Col>
                                <h1>Sorry this page could not be found!</h1>

                                <h2><a href={"/"}>Click here</a> to return to the home page.</h2>
                            </Col>
                        </Row>
                    </Container>
                </Col>
            </Row>
        </Container>
    );

}

export default ErrorPage;
