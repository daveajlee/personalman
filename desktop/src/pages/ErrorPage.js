import React from "react";

const ErrorPage = () => (

    <h1>Error Page - Coming Soon!</h1>

);

export default ErrorPage;
