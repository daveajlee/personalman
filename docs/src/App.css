/* Define the maximum height of the logo image */
.img-max-height {
  max-height:50px;
}

.color-nav {
  background-color: rgb(21,94,111);
}

.color-footer {
  background-color: white;
  margin-top: 10px;
}