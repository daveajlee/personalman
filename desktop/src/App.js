import 'bootstrap/dist/css/bootstrap.min.css';
import './css/App.css'
import Login from "./components/pages/Login/Login";
import RegisterCompany from "./components/pages/RegisterCompany/RegisterCompany";
import React, {useEffect, useState} from "react";
import axios from "axios";
import {Route, Routes} from "react-router-dom";
import RegisterUser from "./components/pages/RegisterUser/RegisterUser";
import AbsenceManagement from "./components/pages/AbsenceManagement/AbsenceManagement";
import ChangePassword from "./components/pages/ChangePassword/ChangePassword";
import UsersManagement from "./components/pages/UsersManagement/UsersManagement";
import AllAbsences from "./components/pages/AllAbsences/AllAbsences";
import Logout from "./components/pages/Logout/Logout";
import './i18n/config';

/**
 * This is the first page that the user visits when starting PersonalMan - either show the login screen or the register
 * a company screen if no companies have been defined so far.
 * @returns {JSX.Element} to be displayed to the user.
 */
function App() {

  const [companies, setCompanies] = useState([]);
  const [loading, setLoading] = useState(true);

    /**
     * Load the current list of companies that are available on the REST API server.
     */
  useEffect(() => {
    axios.get(process.env.REACT_APP_SERVER_URL + `/companies/`)
        .then(res => {
          setLoading(false);
          const companies = res.data;
          setCompanies(companies);
        }).catch(error => {
            console.log('No connection to API');
    })
  }, []);

    /**
     * Display the relevant elements and data to the user.
     */
    return (
        <Routes>
            <Route exact path="/" element={!loading && companies && companies.length > 0 ? <Login/> : <RegisterCompany/>}/>
            <Route path='/login' element={<Login/>} />
            <Route path='/registerCompany' element={<RegisterCompany/>} />
            <Route path='/registerUser' element={<RegisterUser/>} />
            <Route path='/absences' element={<AbsenceManagement/>} />
            <Route path='/changePassword' element={<ChangePassword/>} />
            <Route path='/users' element={<UsersManagement/>} />
            <Route path='/allAbsences' element={<AllAbsences/>} />
            <Route path='/logout' element={<Logout/>} />
        </Routes>
  );
}

export default App;
