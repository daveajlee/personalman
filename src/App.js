import 'bootstrap/dist/css/bootstrap.min.css';
import './css/App.css'
import {Badge} from "react-bootstrap";
import Header from "./components/Header";

function App() {
  return (
    <div class={"container-fluid"}>
      <Header></Header>
      <h1>Welcome to PersonalMan Web Client <Badge bg="secondary">New</Badge></h1>
    </div>
  );
}

export default App;
