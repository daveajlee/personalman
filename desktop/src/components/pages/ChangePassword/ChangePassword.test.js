import renderer from 'react-test-renderer';
import {createBrowserRouter, RouterProvider} from "react-router-dom";
import React from "react";
import ChangePassword from "./ChangePassword";
import App from "../../../App";
import ErrorPage from "../ErrorPage/ErrorPage";

it('checks that the ChangePassword page can be rendered', () => {

    const component = renderer.create(
        <React.StrictMode>
            <RouterProvider router={createBrowserRouter([{
                path: "*",
                element: <App/>,
                errorElement: <ErrorPage />,
            }])}>
            <ChangePassword/>
            </RouterProvider>
        </React.StrictMode>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
