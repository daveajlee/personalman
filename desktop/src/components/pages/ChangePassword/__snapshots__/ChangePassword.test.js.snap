// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checks that the ChangePassword page can be rendered 1`] = `
<div
  className="p-3 my-5 h-custom container-fluid"
>
  <div
    className="d-flex flex-row align-items-center justify-content-center row"
  >
    <div
      className="col"
    >
      <img
        alt="PersonalMan Logo"
        className="d-block mx-auto img-fluid w-50"
        src="https://www.davelee.de/common/assets/img/portfolio/personalman-logo.png"
      />
    </div>
    <div
      className="col"
    >
      <div
        className="d-flex flex-row align-items-center justify-content-center mb-3 container"
      >
        <div
          className="row"
        >
          <div
            className="col"
          >
            <h3>
              Getting Started: Registering a new Company/Organisation
            </h3>
          </div>
        </div>
      </div>
      <div
        className="d-flex flex-column container"
      >
        <form
          className=""
        >
          <div
            className="mb-3 row"
          >
            <label
              className="form-label col-form-label col-sm-2"
              htmlFor="formPlaintextName"
            >
              Name:
            </label>
            <div
              className="col-sm-10"
            >
              <input
                className="form-control"
                id="formPlaintextName"
                onChange={[Function]}
                placeholder="Name"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            className="mb-3 row"
          >
            <label
              className="form-label col-form-label col-sm-2"
              htmlFor="formPlaintextEmail"
            >
              Default Annual Leave:
            </label>
            <div
              className="col-sm-10"
            >
              <input
                className="form-control"
                id="formPlaintextEmail"
                max="40"
                min="20"
                onChange={[Function]}
                placeholder="A number between 20 and 40"
                type="number"
                value={20}
              />
            </div>
          </div>
          <div
            className="mb-3 row"
          >
            <label
              className="form-label col-form-label col-sm-2"
              htmlFor="formPlaintextCompany"
            >
              Base Country:
            </label>
            <div
              className="col-sm-10"
            >
              <select
                aria-label="Company Name"
                className="form-select"
                id="formPlaintextCompany"
                onChange={[Function]}
                value="Germany"
              >
                <option>
                  Germany
                </option>
                <option>
                  England
                </option>
                <option>
                  Scotland
                </option>
              </select>
            </div>
          </div>
        </form>
      </div>
      <div
        className="align-items-center justify-content-center text-md-start mt-4 pt-2 container"
      >
        <div
          className="row"
        >
          <div
            className="text-center col"
          >
            <button
              className="mb-0 px-5 me-2 btn btn-primary btn-lg"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Register
            </button>
            <button
              className="mb-0 px-5 me-2 btn btn-primary btn-lg"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Reset
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
